### ISSUE TITLE
JAX-1 — mace-jax skeleton, neutral-format weight import, and inference forward with torch parity

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1589]
**Depends on:** FM-1 (#1586) · **Blocks:** JAX-2 (#1590)

**Prerequisite:** the Phase 0 characterization suite is committed — goldens, fixtures and the single-source tolerance table under `tests/golden/`, characterization tests under `tests/unit/`. References of the form `P0-N` name that work.

**Context:** mace-jax is the lean, **inference-only** JAX engine of the v1 stack: it evaluates published foundation models converted to the neutral safetensors+JSON format, and nothing else. This ticket builds it end to end — the `packages/mace-jax` skeleton, the neutral-format loader, the `mace_jax.nn` blocks, the physics outputs (E/F/stress/virials via `jax.grad`), and a registered native-jax reference backend — and gates it on **eager fp64 parity against the committed torch goldens**. Inference-only is structural: there is no `train/`, no `run_train`, no training CLI, and no JAX training anywhere in the project (the educational `tutorials/` dir is evaluation/inspection-only, EDU-1 (#1607)). It is also the numeric dividend: forces and stress are *first* derivatives of the energy (legacy computes them in `get_outputs`, `mace/modules/utils.py:270-357`, over the two autograd helpers `compute_forces` `:33-51` and `compute_forces_virials` `:52-82`; here it is one `jax.value_and_grad` over positions and the strain displacement), so **no jax backend ever needs double-backward** and the reference backend needs **no `jax.custom_vjp`** — pure-jax ops get their VJP from autodiff for free. This ticket is where the tensor-generic KernelBackend Protocol (BKD-1 (#1560)) is exercised for real against a second framework (`TensorT = jax.Array`); if the binding fails here, the op-surface premise fails with it. **Accepted deviation (keep):** there is no legacy-JAX E2E harness — parity is measured against the committed torch goldens and the neutral format only, never against the old `ACEsuit/mace-jax` package (which is *not* a frozen oracle; recycling its forward-pass structure is permitted, unlike for the torch package).

**Interface & constraints:**

1. **Supported models (closed v1.0 list).** A model is on the list only if a committed golden exists for that exact artifact. Between them the three models exercise every block variant the energy family publishes, and their architectures are pinned here so architecture closure is checkable before any code exists:

| # | model | architecture (as published) | why | golden source |
|---|---|---|---|---|
| 1 | **MACE-MP-0 small** (`2023-12-10-mace-128-L0`, `mace/calculators/foundations_models.py:93`) | `ScaleShiftMACE`, 3,847,696 params, r_max 6.0, 89 elements, headless; `RealAgnosticResidualInteractionBlock` ×2; Linear + NonLinear readouts; Bessel + polynomial cutoff, no distance transform, no ZBL | smallest published FM, L=0; the Phase 0 golden anchor; first parity target (cheapest to debug) | P0-2 |
| 2 | **MACE-MPA-0 medium** (`:102`; the `mace_mp()` default, `:144-156`) | `ScaleShiftMACE`, 9,063,204 params, r_max 6.0, 89 elements, `heads=['default']`; `RealAgnosticDensityInteractionBlock` then `RealAgnosticDensityResidualInteractionBlock`; Linear + NonLinear readouts; `AgnesiTransform`; **ZBL on** | the model users get by default; the only listed model exercising ZBL, the Agnesi transform, and the density interaction blocks | FM-2 (#1587) (medium-tier golden is an FM-2 (#1587) obligation) |
| 3 | **MACE-OFF23 medium** (`:114`) | `ScaleShiftMACE`, 1,428,368 params, r_max 5.0, 10 elements, headless; `RealAgnosticInteractionBlock` then `RealAgnosticResidualInteractionBlock`; Linear + NonLinear readouts; no transform, no ZBL | organic-chemistry flagship; the non-density interaction pair | P0-2 covers OFF-small; FM-2 (#1587) extends to OFF23-medium |

Two of the three are **headless** — they carry no `heads` attribute at all, and the FM-1 (#1586) artifact records that fact. `mace_jax` treats a headless artifact as single-head; it is not a multi-head artifact and must not be rejected as one.

Explicitly out of v1.0 (each rejection loud, with the failing gate recorded in docs/registry): large tiers and remaining MP aliases; `mh-0`/`mh-1`/MACE-OMOL (multi-head runtime); MACE-MDP (dipole/polarizability observables); MACE-Polar and MACELES (electrostatics — the polar torch port is ELEC-2 (#1592) and its artifact conversion FM-2 (#1587), neither of which buys a jax runtime); the magnetic family (magmom input, magmom observable, `dE/dm`, the SCF fixed point — MAG-1 (#1584), and no published magnetic artifact exists); MACE-ANI-CC (no golden, dropped). Fine-tuned descendants of a supported model (same architecture, new weights) load and are *expected* to work but are not *supported* — verified only for the list.

2. **Feature surface:**

| feature | v1.0 | rule |
|---|---|---|
| observables | `energy`, `node_energy`, `forces`, `stress`, `virials` | forces = `-jax.grad(E, positions)`; stress/virials via the same symmetric strain displacement `mace_core.physics` specifies, implemented in `mace_jax.physics.outputs`; sign conventions pinned by the P0 goldens, never re-derived |
| derivatives legacy has that jax does not | `hessian`, `edge_forces`, `magforces` (`mace/modules/utils.py:270-283` takes `compute_hessian`, `compute_edge_forces`, `compute_magforces`) | out of scope, each rejected by name at load if the artifact declares it — never silently absent from the output dict |
| heads | **single-head runtime only** (a headless artifact counts as single-head) | multi-head artifacts rejected at load with a message pointing at `mace model select-head` |
| dtypes | float64 (**required** — the parity dtype) and float32 | `jax.config.update("jax_enable_x64", True)` happens at load, **owned by `mace_jax`** — never a global the user must remember; per-dtype tolerances from the edit-locked tolerance table, keyed on jnp dtypes; tf32/bf16 out |
| Clebsch-Gordan basis | `reduced` only (`bases = {"reduced"}`) | a `clebsch_gordan_basis: "full"` artifact is **rejected at load, never silently projected** |
| batching | single-structure / naturally-batched **eager** evaluation | no jit requirement; `jax.jit` may be applied opportunistically but no acceptance criterion depends on it — the padded static-shape jit pipeline is JAX-2 (#1590) and must not leak into this ticket |
| weight loading | **only** the neutral safetensors+JSON format | never a legacy pickle; `mace_jax` **never imports `mace/` or `mace_torch`** (imports `mace_core` only, import-linter enforced); descriptors are rebuilt from the `CanonicalSpec` sidecar and handed to `mace_jax` backend factories |
| training, fine-tuning, LAMMPS/MLIAP, Hessians, dipoles/polarizability, magnetism, electrostatics, committee calculators | **out** | structural, not deferred polish |

3. **Load boundary** (`mace_jax.deploy.neutral_io`) and model registry (`mace_jax.models.registry`):

```python
@dataclass(frozen=True)
class JaxModelSpec:
    model_id: str                    # registry name, e.g. "mace-mpa-0-medium"
    artifact_ref: str                # neutral-format artifact (FM-3 (#1588) resolves to URL + DOI, FM-4 (#1636) fetches it)
    golden_ref: str                  # committed golden id (tests/golden/references/...)
    required_blocks: frozenset[str]  # informative; authoritative source is the sidecar

SUPPORTED_MODELS: dict[str, JaxModelSpec]

def load_model(
    path_or_id: str, *, backend: str = "reference", dtype: str = "float64",
) -> tuple["ApplyFn", "Params"]:
    """Parse JSON sidecar -> descriptors -> mace_jax backend factories -> pure apply fn.
    ApplyFn: (Params, graph: dict[str, jax.Array]) -> MACEOutputs[jax.Array]  # the flat-dict
    graph schema (DATA-3 (#1571)); no typed GraphBatch container at the model boundary.
    Rejects loudly: multi-head runtime, clebsch_gordan_basis != 'reduced',
    unsupported derivative targets, missing blocks (with the list of missing block names)."""
```

Every rejection is loud and actionable: the loader enumerates descriptor kinds from the sidecar and fails with the *list of missing blocks*, so architecture closure is checkable without running anything. The silent-substitution defect class is not reintroduced.

4. **Op surface.** `mace_jax.kernels` binds the framework-generic KernelBackend Protocol from `mace_core.kernels` (BKD-1 (#1560)) to `jax.Array` and ships one **reference jax backend**, registered under the entry-point group `mace.kernel_backends.jax` (`reference = ...`) and discovered by `mace_core.kernels.registry.available_backends("jax")`. Contract-relevant rules:
   - The reference backend is **native jax — no `e3nn-jax`** (e3nn/e3nn-jax are removed project-wide). Tensor products and the symmetric contraction run over `mace_core`'s irreps and the reduced Clebsch-Gordan basis (CORE-4 (#1558)), exactly as the torch reference is native plain torch. Legacy reaches its reduced basis only through `cuequivariance` (`mace/tools/cg.py:106-140`, gated on `CUET_AVAILABLE`), which is CUDA-only and has no jax equivalent in scope — the CORE-4 (#1558) basis is what makes a jax reduced-basis runtime possible at all. `cuequivariance_jax` is the anticipated optional second backend, **post-v1.0**; the entry-point group means adding it is zero core edits.
   - `segment_reduce` maps to `jax.ops.segment_sum` with `num_segments` supplied from the descriptor/graph — **never inferred from `edge_index.max()`** (empty-neighbourhood correctness plus no host sync).
   - Reference-only ops (`spherical_harmonics`, `radial_basis`) plus the model glue that is *not* ops — radial MLP, ZBL pair repulsion, Agnesi distance transform, scale-shift and E0 assembly (per the E0Spec semantics, ARCH-3 (#1562), with the fp64 accumulation invariant expressed through `PrecisionConfig`, not a hardcoded cast) — are plain jax in `mace_jax.nn`, scoped by exactly what the model list needs. Legacy registers **eight** interaction classes and **six** readout classes; the three supported models need exactly four interaction variants (`RealAgnostic{,Residual,Density,DensityResidual}InteractionBlock`, `mace/modules/blocks.py:728,831,934,1055`), two readouts (`LinearReadoutBlock` `:64`, `NonLinearReadoutBlock` `:86`), `BesselBasis` (`mace/modules/radial.py:18`), `PolynomialCutoff` (`:113`), `AgnesiTransform` (`:225`) and `ZBLBasis` (`:149`) — **and nothing else**. `SoftTransform`, `ChebychevBasis`, `GaussianBasis` and the four unused interaction/readout variants are not ported.
   - **No `jax.custom_vjp` anywhere in the reference backend**; the double-backward capability flag is `False` across the whole package by construction (a future fused backend supplies its own VJP).

5. **Module/parameter library — an implementation decision this ticket closes** (left open deliberately; the op surface and loader are pinned, the module system is not). Recommendation: **`flax` (NNX)** — mainstream JAX stack (flax + optax + orbax), what `ACEsuit/mace-jax` already uses; chosen for future training optionality with no rewrite. Lean runner-up: **`equinox`** (modules-as-pytrees, tidiest for injecting canonical weights) — the pick if inference-only is committed forever; it also trains fine (equinox + optax), so this is an ecosystem call, not a capability one. `flax`-linen is not an option (maintenance mode). Low-stakes by construction: the equivariant math is carried functionally by `mace_core`, so the module library is thin plumbing (linears, holding weights, wiring). Record the choice and its rationale in the PR description.

6. **Package skeleton** (`packages/mace-jax`, import name `mace_jax`; deps: jax + `mace_core` only): `graph.py` (flat-dict graph contract bound to `jax.Array`, same schema as `mace_core.graph`), `kernels/{ops.py,reference.py}`, `nn/` (interaction, product basis, symmetric contraction, readout, radial), `models/{base.py,energy.py}` (BaseMACE-jax → `MACEOutputs`; MACE/ScaleShiftMACE-jax), `physics/outputs.py`, `deploy/neutral_io.py`, `cli/eval.py` (inference-only), `tests/{test_parity_neutral.py,test_equivariance.py}`. **No `train/`, no `cli/run_train.py`.**

7. **Conformance entry point.** `run_jax_model_conformance(model_id, *, dtypes=("float64","float32"), devices=("cpu",), tol=None)` runs golden parity per dtype + equivariance for one supported model; it is what CI calls. CI tiering (decided, wire only what this ticket needs): a per-PR **CPU tiny-anchor parity job, Required, no downloads** (the committed P0-1 artifact), added to `ci-core.yaml` alongside `lint`/`unit`/`workflows`/`coverage` and driven through the `.github/actions/run-tests` composite like every other job; the nightly full-list rows (download converted artifacts, CPU + NVIDIA) hang off this same entry point via the FM-3 (#1588) badge matrix. The jit-stability/recompile-count gate belongs to JAX-2 (#1590) and must not be back-ported into this ticket's acceptance.

**Task:**
1. Create `packages/mace-jax` per the skeleton above (no training surface anywhere); wire the import-linter contract: `mace_jax` imports `mace_core` only — never `mace/`, never `mace_torch`, never e3nn-jax.
2. Decide and record the module library (flax NNX recommended, equinox runner-up); implement `mace_jax.nn` for the exact block list in constraint 4 and `mace_jax.models.energy` (MACE/ScaleShiftMACE-jax → `MACEOutputs`).
3. Implement `mace_jax.kernels`: the Protocol bound to `jax.Array`; the native-jax reference backend (TP/contraction over `mace_core`'s reduced Clebsch-Gordan basis; `segment_sum` with static `num_segments`); register it under `mace.kernel_backends.jax` and confirm discovery via `available_backends("jax")`.
4. Implement `mace_jax.physics.outputs`: energy, node_energy, forces (`-jax.grad`), stress/virials via the symmetric strain displacement per `mace_core.physics`; fp64 enabled at load by `mace_jax` itself.
5. Implement `mace_jax.deploy.neutral_io.load_model` with the four loud rejections (multi-head → `mace model select-head` pointer; full basis; unsupported derivative target; missing blocks with the list) and `SUPPORTED_MODELS` with the three `JaxModelSpec` rows.
6. Parity tests: fp64 E/F/stress vs the committed torch goldens for the tiny anchor (P0-1) and the three supported models (converted via the FM-1 (#1586) pipeline; medium-tier goldens per FM-2 (#1587)), within the edit-locked per-dtype tolerance table; equivariance tests; fp32 rows at fp32 tolerance.
7. Implement `run_jax_model_conformance` and wire the per-PR CPU tiny-anchor job in `ci-core.yaml` (paths-filtered on `packages/mace-jax/**` + `mace_core` kernel/Clebsch-Gordan paths, using the `dorny/paths-filter` pattern `ci-extensions.yaml` already establishes; no network).

**Out of scope:** the padded static-shape jit pipeline, recompile-count gate, benchmark, and JAX ASE calculator (JAX-2 (#1590)); jax-md and LAMMPS-from-jax (deferred with recorded reopening conditions); the `cuequivariance_jax` backend (post-v1.0); multi-head runtime, MDP/Polar/LES/magnetic/OMOL/large-tier support (each fails a recorded gate); any training or fine-tuning code; artifact download and caching (FM-4 (#1636)); the nightly badge matrix itself (FM-3 (#1588)).

**Acceptance criteria:**
- [ ] `packages/mace-jax` installs alongside the other packages; it ships no `train/` and no training CLI; import-linter proves `mace_jax` imports `mace_core` only (no `mace/`, no `mace_torch`, no e3nn-jax).
- [ ] `load_model` reads the neutral safetensors+JSON format only and rejects loudly: multi-head (with the `mace model select-head` pointer), `clebsch_gordan_basis != "reduced"` (never silently projected), an artifact declaring `hessian`/`edge_forces`/`magforces`, and missing blocks (with the missing-block list). A **headless** artifact loads as single-head.
- [ ] The reference jax backend is registered under `mace.kernel_backends.jax` and discovered by `mace_core.kernels.registry.available_backends("jax")`; it contains no `jax.custom_vjp` and no e3nn-jax; `segment_reduce` never reads `edge_index.max()`.
- [ ] `mace_jax.nn` implements exactly the four interaction variants, two readouts, Bessel, polynomial cutoff, Agnesi and ZBL the three models need, and nothing else — asserted against the block list, so an unused legacy variant cannot creep in.
- [ ] fp64 E/F/stress parity vs the committed torch goldens passes for the tiny anchor and all three supported models within the edit-locked tolerance table; fp32 passes at its own tolerance row; equivariance tests pass.
- [ ] fp64 (`jax_enable_x64`) is enabled by `mace_jax` at load — a fresh process with no user-set globals produces the fp64 parity numbers.
- [ ] `run_jax_model_conformance` exists and is what CI calls; the per-PR CPU tiny-anchor parity job is green with no network access.
- [ ] No acceptance test depends on `jax.jit`; eager evaluation is the contract.

**Verify:**
```bash
python -m pytest packages/mace-jax/tests -v                              # unit, equivariance, tiny-anchor parity (fp64+fp32, no network)
python -m pytest packages/mace-jax/tests/test_parity_neutral.py -v       # neutral-format load -> goldens
lint-imports                                                              # mace_jax -> mace_core only
MACE_CI_ALLOW_NETWORK=1 python -m pytest packages/mace-jax/tests -m network -v   # full three-model conformance (downloads converted artifacts)
```
**Review focus:** the parity evidence (all three models, both dtypes, against committed goldens — not weight equality); that the loader's four rejections are loud and tested and that headless is not mistaken for multi-head; that nothing in the package needs double-backward or `custom_vjp`; and that no JAX-2 (#1590) content (padding, jit gates, ASE calculator) leaked into acceptance. Physics reviewer on the stress/virial sign conventions vs the P0 goldens.

---
